frame/nft-fractionalization/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+		_,
+		Blake2_128Concat,
+		(T::NftCollectionId, T::NftId),
+		Details<AssetIdOf<T>, AssetBalanceOf<T>>,
+		OptionQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An NFT was successfully fractionalized.
+		NftFractionalized {
+			nft_collection: T::NftCollectionId,
+			nft: T::NftId,
+			fractions: AssetBalanceOf<T>,
+			asset: AssetIdOf<T>,
+			beneficiary: T::AccountId,
+		},
+		/// An NFT was successfully returned back.
+		NftUnified {
+			nft_collection: T::NftCollectionId,
+			nft: T::NftId,
+			asset: AssetIdOf<T>,
+			beneficiary: T::AccountId,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Information about the fractionalized NFT can't be found.
+		DataNotFound,
+		/// The signing account has no permission to do the operation.
+		NoPermission,
+		/// NFT doesn't exist.
+		NftNotFound,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::fractionalize())]
+		pub fn fractionalize(
+			origin: OriginFor<T>,
+			nft_collection_id: T::NftCollectionId,
+			nft_id: T::NftId,
+			asset_id: AssetIdOf<T>,
+			beneficiary: AccountIdLookupOf<T>,
+			fractions: AssetBalanceOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let beneficiary = T::Lookup::lookup(beneficiary)?;
+
+			let nft_owner =
+				T::Nfts::owner(&nft_collection_id, &nft_id).ok_or(Error::<T>::NftNotFound)?;
+			ensure!(nft_owner == who, Error::<T>::NoPermission);
+
+			let pallet_account = Self::get_pallet_account();
+			Self::do_lock_nft(nft_collection_id, nft_id)?;
+			Self::do_create_asset(asset_id, pallet_account.clone())?;
+			Self::do_mint_asset(asset_id, &beneficiary, fractions)?;
+			Self::do_set_metadata(asset_id, &who, &pallet_account, &nft_collection_id, &nft_id)?;
+
+			NftToAsset::<T>::insert(
+				(nft_collection_id, nft_id),
+				Details { asset: asset_id, fractions },
+			);
+
+			Self::deposit_event(Event::NftFractionalized {
+				nft_collection: nft_collection_id,
+				nft: nft_id,
+				fractions,
+				asset: asset_id,
+				beneficiary,
+			});
+
+			Ok(())
+		}
+
+		/// Burn the whole amount of the asset and return back the locked NFT.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::unify())]
+		pub fn unify(
+			origin: OriginFor<T>,
+			nft_collection_id: T::NftCollectionId,
+			nft_id: T::NftId,
+			asset_id: AssetIdOf<T>,
+			beneficiary: AccountIdLookupOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let beneficiary = T::Lookup::lookup(beneficiary)?;
+
+			NftToAsset::<T>::try_mutate_exists((nft_collection_id, nft_id), |maybe_details| {
+				let details = maybe_details.take().ok_or(Error::<T>::DataNotFound)?;
+				ensure!(details.asset == asset_id, Error::<T>::DataNotFound);
+
+				Self::do_burn_asset(asset_id, &who, details.fractions)?;
+				Self::do_unlock_nft(nft_collection_id, nft_id, &beneficiary)?;
```

Comment:
i wonder if we can add more specific Error types for these calls.

Reply:
You'll get the real errors from the underlying methods if something goes wrong, as I remember they're pretty specific